### PR TITLE
[libc++][hardening] Classify assertions related to leaks and syscalls.

### PR DESCRIPTION
**Pull request by var-const**
Introduce two new categories:
- `_LIBCPP_ASSERT_VALID_DEALLOCATION`;
- `_LIBCPP_ASSERT_VALID_EXTERNAL_API_CALL`.

### CODE REVIEW

File: libcxx/src/filesystem/operations.cpp
Comment (var-const) [inline libcxx/src/filesystem/operations.cpp:673]: Something I wanted to point out about the new `valid-api-call` category is that this is (so far, at least) the only category that also covers spurious failures. All other categories are (or are supposed to be) completely deterministic -- they are direct consequences of the user giving us invalid arguments. System calls, however, can (at least in theory) fail sporadically, with no fault on the user's part.

I'd like to make sure we think it's reasonable. Alternatively, we could:
- only leave checks that we only expect to fail due to bad user input that is hard to check. A failure to lock a mutex might be a good example -- it's most likely to happen if the user didn't lock the mutex in the first place (but we don't have an easy way to check for that, so we have to settle for checking the consequences). However, even this category still might cover spurious errors;
- separate the user-triggered failures and the spurious failures. For the latter, either create a new category or mark them `internal`. The downsides are a) arguably too fine a granularity; b) the separation between what is triggered by a user and what is spurious isn't completely deterministic.

File: libcxx/src/filesystem/operations.cpp
Comment (hawkinsw) [reply libcxx/src/filesystem/operations.cpp:464]: Here is something interesting that I found from the `man` page:

```
       •  If name corresponds to a maximum or minimum limit, and that limit is
          indeterminate, -1 is returned and errno is not changed.  (To distin‐
          guish an indeterminate limit from an error, set errno to zero before
          the call, and then check whether errno is nonzero  when  -1  is  re‐
          turned.)
```

So, perhaps, there could be a platform out there that has an indeterminate maximum path length that could return `-1` and mean that it was not an error?

File: libcxx/src/filesystem/operations.cpp
Comment (hawkinsw) [reply libcxx/src/filesystem/operations.cpp:624]: It *seems* possible -- you could have SELinux enabled that allows the owner of the process to `read` the file but denies it `getattr`?

File: libcxx/include/__config
Comment (mordante) [inline libcxx/include/__config:290]: Not specific to this commit, but I feel this list is getting a bit long in here. How about putting it in an `.rst` file. I also would like a better overview what the modi do. For example, what does `_LIBCPP_HARDENING_MODE_FAST`  do.
Maybe a summary lke
```
* _LIBCPP_HARDENING_MODE_FAST
  * foo
  * bar
* _LIBCPP_HARDENING_MODE_EXTENSIVE
   * all of _LIBCPP_HARDENING_MODE_FAST
   * do
   * re
   * mi
```

File: libcxx/include/__coroutine/coroutine_handle.h
Comment (mordante) [reply libcxx/include/__coroutine/coroutine_handle.h:58]: I agree it might look a bit odd, but you could call it an external API. I don't mind to keep it in the same category, but I would like to specifically mention that compiler intrinsics are part of this category.

File: libcxx/src/filesystem/operations.cpp
Comment (mordante) [reply libcxx/src/filesystem/operations.cpp:673]: I also wonder how often these can trigger when an end-user deletes a file the application was using. AFAIK we don't enable these assertions in the dylib so at the moment they are unchecked. Right?

I also how many of these can be solved by throwing an exception. Obviously that requires validating the functions against the WP.

File: libcxx/include/__coroutine/coroutine_handle.h
Comment (ldionne) [reply libcxx/include/__coroutine/coroutine_handle.h:58]: I agree, I think I would mention that we consider anything external to the library (including compiler builtins) as an "external API" for the purpose of this category.

File: libcxx/src/filesystem/operations.cpp
Comment (hawkinsw) [reply libcxx/src/filesystem/operations.cpp:464]: @ldionne Your reading seems to match with the reading that I had from the `man` page and the section that I quoted! I am glad that we are seeing the same thing!

File: libcxx/include/__coroutine/coroutine_handle.h
Comment (var-const) [reply libcxx/include/__coroutine/coroutine_handle.h:58]: Good points, done.

File: libcxx/src/filesystem/operations.cpp
Comment (var-const) [reply libcxx/src/filesystem/operations.cpp:624]: Added a TODO to revisit this later.

File: libcxx/src/filesystem/operations.cpp
Comment (var-const) [reply libcxx/src/filesystem/operations.cpp:464]: I removed the assertion in favor of error reporting. There's an additional quirk that apparently `PATH_MAX` might not be available on z/OS which I presume false under the `#else` branch of the `#if defined(_LIBCPP_WIN32API) || defined(__GLIBC__) || defined(__APPLE__)` condition (it really comes down to the `__GLIBC__` define -- I don't know if lack of `glibc` can be used as a reliable indicator of z/OS).

File: libcxx/include/__config
Comment (var-const) [reply libcxx/include/__config:290]: I completely agree, and I plan to improve the documentation after this round of categorization patches lands.

File: libcxx/src/filesystem/operations.cpp
Comment (var-const) [reply libcxx/src/filesystem/operations.cpp:673]: Removed the assertion.

File: libcxx/include/__config
Comment (mordante) [reply libcxx/include/__config:302]: That probably happens after branching LLVM 18. After branching we can always backport documentation changes. So it would be great if we can get better documentation in LLVM 18.

File: libcxx/test/libcxx/utilities/utility/mem.res/mem.poly.allocator.class/mem.poly.allocator.mem/assert.deallocate.pass.cpp
Comment (mordante) [inline libcxx/test/libcxx/utilities/utility/mem.res/mem.poly.allocator.class/mem.poly.allocator.mem/assert.deallocate.pass.cpp:2]: This changes seems unintended.

**Review by mordante — APPROVED** [review]
LGTM modulo one nit.

**Review by ldionne — APPROVED** [review]
LGTM.

File: libcxx/test/libcxx/utilities/utility/mem.res/mem.poly.allocator.class/mem.poly.allocator.mem/assert.deallocate.pass.cpp
Comment (var-const) [reply libcxx/test/libcxx/utilities/utility/mem.res/mem.poly.allocator.class/mem.poly.allocator.mem/assert.deallocate.pass.cpp]: Thanks!

**Pull request merged**
